### PR TITLE
Adding new Battery formulations

### PR DESCRIPTION
**Pull request by sourabhdalvi**
New Formulations
- [x] BookKeeping with end of period energy target (hard) constraint 
- [x] BookKeeping with end of period energy target (soft) constraint with slack variable
- [x]  BookKeeping with end of period value of energy

Requires PSY#685 to be merged

### CODE REVIEW

File: src/devices_models/devices/storage.jl [hunk 1]
```diff
@@ -3,7 +3,11 @@
 abstract type AbstractStorageFormulation <: AbstractDeviceFormulation end
 struct BookKeeping <: AbstractStorageFormulation end
 struct BookKeepingwReservation <: AbstractStorageFormulation end
-
+struct BookKeepingwTarget <: AbstractStorageFormulation end
+struct BookKeepingwTargetTimeSeries <: AbstractStorageFormulation end
+struct BookKeepingwSoftTarget <: AbstractStorageFormulation end
+struct BookKeepingwSoftTargetTimeSeries <: AbstractStorageFormulation end
+struct BookKeepingwEnergyValue <: AbstractStorageFormulation end
 ########################### ActivePowerInVariable, Storage #################################
```
Comment (claytonpbarrows) [inline src/devices_models/devices/storage.jl:11]: abstract type BookKeeping <: AbstractStorageFormulation
abstract type StorageTarget <: BookKeeping
struct StorageTargetTimeSeries <: StorageTarget

File: src/devices_models/devices/storage.jl [hunk 2]
```diff
@@ -335,64 +347,66 @@ end
...
+                UpdateRef{St}(TARGET, target_forecast_label),
+            )
+        else
+            energy_soft_target_timeseries(
+                psi_container,
+                constraint_infos_target,
+                make_constraint_name(ENERGY_TARGET, St),
+                (
+                    make_variable_name(ENERGY, St),
+                    make_variable_name(ENERGY_TARGET_SLACK, St),
```
Comment (claytonpbarrows) [inline src/devices_models/devices/storage.jl:389]: I had previously recognized a pattern that all `*._SLACK` variables were subject to a `*._SLACK_COST` defined in https://github.com/NREL-SIIP/PowerSimulations.jl/blob/sd/battery_model/src/core/definitions.jl. This breaks that convention and makes this slack subject to the cost defined in the struct. 
I'm not necessarily saying that we should keep the old convention and change this terminology, but rather highlighting the change and asking if it matters to anyone (@jd-lara)?

Comment (sourabhdalvi) [reply src/devices_models/devices/storage.jl:389]: `ENERGY_TARGET_SHORTAGE`/`ENERGY_SHORTAGE`?

File: src/devices_models/devices/storage.jl [hunk 3]
```diff
@@ -1,9 +1,16 @@
 #! format: off
 
 abstract type AbstractStorageFormulation <: AbstractDeviceFormulation end
+abstract type EnergyManagementFormulation <: AbstractStorageFormulation end
+abstract type EnergyTargetFormulation <: EnergyManagementFormulation end
+abstract type RelaxedEnergyTargetFormulation <: EnergyManagementFormulation end
 struct BookKeeping <: AbstractStorageFormulation end
 struct BookKeepingwReservation <: AbstractStorageFormulation end
-
+struct EndOfPeriodStaticEnergyTarget <: EnergyTargetFormulation end
```
Comment (jd-lara) [inline src/devices_models/devices/storage.jl:9]: We need docstrings for these formulations. It isn't at all clear what the model is from the description.

**Review by jd-lara — COMMENTED** [review]
I am not totally convinced that the formulation abstractions is adequate since the formulations seem to be basically the same except for the parameters.

**Pull request closed without merging**